[SCHEMA] Add metadata term files (#774)
bids-standard/bids-specification: src/04-modality-specific-files/01-magnetic-resonance-imaging-data.md

### Scanner Hardware

{{ MACROS___make_metadata_table(
   {
      "Manufacturer": "RECOMMENDED",
      "ManufacturersModelName": "RECOMMENDED",
      "DeviceSerialNumber": "RECOMMENDED",
      "StationName": "RECOMMENDED",
      "SoftwareVersions": "RECOMMENDED",
      "HardcopyDeviceSoftwareVersion": "DEPRECATED",
      "MagneticFieldStrength": "RECOMMENDED, but REQUIRED for Arterial Spin Labeling",
      "ReceiveCoilName": "RECOMMENDED",

>       "ReceiveCoilActiveElements": "RECOMMENDED",

Reviewer: I think this is how you add on to the description?

```suggestion
      "ReceiveCoilActiveElements": {"RECOMMENDED", "See an example below the table."}
```

Author: The code checks for a tuple instead of a set, but other than that it's right.

### In-Plane Spatial Encoding

{{ MACROS___make_metadata_table(
   {
      "NumberShots": "RECOMMENDED",
      "ParallelReductionFactorInPlane": "RECOMMENDED",
      "ParallelAcquisitionTechnique": "RECOMMENDED",
      "PartialFourier": "RECOMMENDED",
      "PartialFourierDirection": "RECOMMENDED",
      "PhaseEncodingDirection": (
         "RECOMMENDED",
         "This parameter is REQUIRED if corresponding fieldmap data is present "
         "or when using multiple runs with different phase encoding directions "
         "(which can be later used for field inhomogeneity correction).",
      ),
      "EffectiveEchoSpacing": (
         "RECOMMENDED",

>          "This parameter is REQUIRED if corresponding fieldmap data is present.",

Reviewer: ```suggestion
         "<sup>2</sup> This parameter is REQUIRED if corresponding fieldmap data is present.",
```

      ),
      "TotalReadoutTime": (
         "RECOMMENDED",

>          "This parameter is REQUIRED if corresponding 'field/distortion' maps "
>          "acquired with opposing phase encoding directions are present "
>          "(see 8.9.4).",

Reviewer: Add superscript and also finally update this out-of-date reference to a section number.

```suggestion
         "<sup>3</sup> This parameter is REQUIRED if corresponding 'field/distortion' maps "
         "acquired with opposing phase encoding directions are present "
         "(see [Case 4: Multiple phase encoded "
         "directions](#case-4-multiple-phase-encoded-directions-pepolar)).",
```

### Required fields

{{ MACROS___make_metadata_table(
   {
      "RepetitionTime": "REQUIRED",

>       "VolumeTiming": "REQUIRED",

Reviewer 2: Wouldn't it make sense to say:

```suggestion
      "VolumeTiming": "REQUIRED, unless `RepetitionTime` and `DelayTime` are defined.",
```

And make similar adjustments for RepetitionTime and DelayTime requirement levels?

Reviewer 2: I notice that there are more fields that could benefit from such an overhaul (AcquisitionDuration)

Author: Could we push this one to a separate issue/PR? I think it ties into that weird dependency table for timing-related field and will require more thought than I can put in at the moment.

![image](https://user-images.githubusercontent.com/8228902/124508108-3db7ed00-dd9d-11eb-948c-c001c72d7794.png)

Reviewer 2: ugh, you are right. Let's not open that can of worms (especially not me).

#### Timing Parameters

{{ MACROS___make_metadata_table(
   {
      "NumberOfVolumesDiscardedByScanner": "RECOMMENDED",
      "NumberOfVolumesDiscardedByUser": "RECOMMENDED",
      "DelayTime": "RECOMMENDED",

>       "AcquisitionDuration": 'RECOMMENDED, but REQUIRED for sequences that are described with the `VolumeTiming` field and that do not have the `SliceTiming` field set to allow for accurate calculation of "acquisition time"',

Reviewer 2: the linebreaks in the table rendering differ for this field in the spec, with the new version being a bit worse, see old version screenshot:

![image](https://user-images.githubusercontent.com/9084751/124444868-531e1e80-dd7f-11eb-97a9-e6d9ebcd40fb.png)

new version (a bit worse)

![image](https://user-images.githubusercontent.com/9084751/124444915-5dd8b380-dd7f-11eb-898b-081d6ab3a8ce.png)

It's minor, but perhaps you easily see what can be changed to have this work better. If not, no worries.

Author: I think other changes are probably just making that column slightly narrower overall. I don't see anything in that string that would explain the change.

Reviewer 2: Alright, thanks for checking 👍

Fieldmap data MAY be linked to the specific scan(s) it was acquired for by
filling the `IntendedFor` field in the corresponding JSON file.

{{ MACROS___make_metadata_table(
   {
      "IntendedFor": (

>          "RECOMMENDED",

Reviewer: ```suggestion
         "OPTIONAL",
```

Required fields:

{{ MACROS___make_metadata_table(
   {

>       "EchoTime": "REQUIRED",

Reviewer: This needs to just be a number, not a number or array of numbers.
![Screenshot from 2021-04-28 12-34-23](https://user-images.githubusercontent.com/83442/116440023-146b6580-a81e-11eb-9720-91dd4c0e24da.png)

Author: Maybe we can include a caveat in the description here? EchoTime in the general sense can be a number or array of numbers, but obviously that's not really applicable to field maps.

Reviewer: Yeah, I'm not sure what the best move is here. It feels weird to say "Note that much of the preceding does not apply here." What if we made this `"fmapEchoTime": "REQUIRED"` with a modified type and description?

Author: Wouldn't that be introducing a new field? What about updating the base definition to say something like,

> This field MAY be an array when it references a 4D imaging file in which volume timing is critical for interpretation, such as in [ASL](/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#arterial-spin-labeling-perfusion-data) or variable echo time fMRI sequences.

Reviewer: I assumed `name` was what was going into the field, not the filename... This makes it a bit harder to customize.

The problem with changing that last bit of text is that it was pretty heavily workshopped during the ASL/qMRI discussions to satisfy everybody.

Author: >I assumed name was what was going into the field, not the filename... This makes it a bit harder to customize.

I honestly haven't leveraged `name` at all so far. I don't want to rely too heavily on `name` and the filename not matching, but it could be very useful in special cases, which this might count as.

>The problem with changing that last bit of text is that it was pretty heavily workshopped during the ASL/qMRI discussions to satisfy everybody.

Good point.

Author: I've updated the code to do as you requested, and it should be good as of 3e2a56d.
